io/include/detray/io/common/detail/type_traits.hpp

```diff
@@ -18,6 +21,107 @@
 
 namespace detray::detail {
 
+/// Determine the type and id of a shape of a mask without triggering a compiler
```

Comment:
Could you explain the changes in this file? 
It is not going to be nice if we have to add all mask information for all mask classes.
I hope this information can be moved into the mask classes or look for another way

Reply:
At one point I need to match the shape ids that are used during the building to the specific mask ids that a given detector needs and, as far as I can tell, this has be done for all mask classes that we want to distinguish in io payloads.

However, not all detectors will have all mask types or ```enum```s for these mask types defined. E.g. if you add a new type yourself, you should not have to refactor all other detectors in your build to include the same type/enum and keep it empty.  So, for example, I would expect that
```
auto mask_id{detector_t::masks::id::e_somebodys_funky_shape}
```
will not compile if ```detector_t``` is the toy detector, but I have not checked this.
But this new type needs to be addressed in the building code somehow when your particular detector is being built.  I am pretty sure this would even happen in a pattern like
```
if constexpr (std::is_same_v<detector_t, somebodys_detector_t>) {
     auto mask_id{detector_t::masks::id::e_somebodys_funky_shape}
}
```
However, if this faulty access happens during template substitution it will not trigger an error (sfinae) and that is why these structs  are done using template specialization: The mask info type for a non existing shape will not be compiled once the substitution is wrong, defaulting to the shape type ```void``` for the particular ```detector_t```

Reply:
Not sure if I am following your explanation.

Let's say that a random user wants to build a detector with vertex strip and drift chambe.
The user will need to use `rectangle2D<1, true>` and `line<true, 1, true>` which is not defined in `type_traits.hpp` nor `detector_metadata.hpp`

Do that random user have to edit both files?

Reply:
Yes and no. The user should write a detector metadata for their specific detector anyway, and there they need to add exactly those types they need. Any previously unknown type will then have to be added to the IO, where it will need an io mask shape ID and this matching of the shape id to the detector mask id, but the example ```detector_metadata.hpp``` in detray does not need to be touched.

Reply:
The detray core types we should add however

Reply:
Hmm, can we define a unique eigen number for every mask type (defined with the template parameters of mask type) as we do for the geometry id and use it over the entire project including metadata and io?

Reply:
I am not trying to fix the codes right now (I understand the situation). We can proceed as it is for the moment. 
I just wanted to discuss whether we can improve the code later for better maintainability in the future
It is hard to know if it is possible or even beneficial before I start working on it.

Reply:
Yes, no objections there, but let's not do that right away, I would like to finish an initial implementation first :P
This code is anything from final, since the way the material and grids get decorated is not solved in this PR. The matching of the grid types could become much more complicated than this and maybe I can come up with something smarter then, but I thought it would help to try and do things in smaller steps for once :D

Reply:
> Yes, no objections there, but let's not do that right away

OK. It is enough to know